### PR TITLE
Mark non-final microbatches for MFSDP v2 hybrid data parallelism

### PR DESCRIPTION
**Pull request by wujingyue**
## Problem

MFSDP v2's hybrid data parallelism drops accumulated gradients: with multiple
microbatches, only the last microbatch's gradient reaches the optimizer.
`reduce_partial_gradients` uses `context.is_last_microbatch` to defer outer-DP
reduction, but the adapter inherited a no-op `no_sync`, leaving that flag true
for every backward. Finalizing the accumulation buffer on every backward caused
the next microbatch to clear it.

## Changes

- Enter `microbatch(is_last=False)` from `FullyShardedDataParallelV2.no_sync`.
  Install the v2 `no_sync_func` regardless of `overlap_grad_reduce`, because v2
  reduces gradients during backward in either mode.
- Require the `is_last_microbatch` keyword at the gradient-reduction call site.
- Add accumulation regression coverage and three functional cases with new
  H100 dev golden values for NGC 26.08: baseline, HSDP, and HFSDP.
- Explicitly set `expert_outer_dp_sharding_strategy='optim'` in the HFSDP
  recipe and `no_shard` in the baseline and HSDP recipes. This PR preserves
  the current config and CLI defaults and does not depend on a default-inheritance change.

## Validation

Rebased onto `main` at `f6c33bde4`; no merge commit.

- The eight-H100 hybrid unit suite passed before the default-inheritance
  change was split into a separate PR. This PR retains the
  gradient-accumulation regression coverage.
- Black, isort, and Ruff checks passed for the changed Core/test files.
- All three functional cases passed in independent fresh 50-iteration runs on
  8 x H100 with NGC 26.08: `deepseek_proxy_mfsdp_v2_ep2_mb2`,
  `deepseek_proxy_mfsdp_v2_ep2_hsdp_mb2`, and `deepseek_proxy_mfsdp_v2_ep2_hfsdp_mb2`.
  Each passed exact loss and approximate memory checks against its new goldens.
  Iteration-1 gradient norm was `10.681` for all three.
  The recipes now explicitly select the same effective expert outer strategies
  used in those runs; training was not rerun after that mechanical change.

Exact loss checks remain enabled. Approximate loss tolerance alone does not
reliably detect the original accumulation bug, so these cases must not adopt
`NON_DETERMINSTIC_RESULTS=1`. Each case checks its own golden; the functional
harness does not compare the three configurations directly.

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py [hunk 1]
```diff
@@ -366,7 +366,7 @@ def _has_sharded_grads(self) -> bool:
         return has_any_grad
 
     def reduce_partial_gradients(
-        self, partial_grad: DBuffer, is_last_microbatch: bool = True
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/parameter_group.py:408]: `is_last_microbatch` is important enough to be required.

File: megatron/core/distributed/fsdp/mcore_fsdp_adapter.py [hunk 2]
```diff
@@ -812,6 +814,21 @@ def _validate_config(
...
+        HSDP/HFSDP leave the DP-outer axis Partial across microbatches and reduce it
+        on the last backward of a step, so MFSDP has to be told which backward that
+        is. Without it every backward finalizes that axis and marks the accumulation
+        buffer stale, so the next microbatch zeroes it and only the last microbatch's
+        gradient reaches the optimizer.
+
+        MCore's schedules wrap every microbatch but the last in ``no_sync_func``.
+        """
+        with microbatch(self.module.context, is_last=False):
+            yield
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/mcore_fsdp_adapter.py:828]: Neat!

**Review by shjwudp — APPROVED** [review]

**Review by ko3n1g — APPROVED** [review]

**Review by maanug-nv — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]
